src/platform-aws.c

```diff
@@ -759,63 +759,141 @@ static int get_rail_vf_idx(struct fi_info *info)
 	return vf_idx;
 }
 
+
+/*
+ * On P5/P5e, there are up to 32 EFA devices.  Each pair of EFA
+ * devices shares some Nitro card resources, and there is a marginal
+ * performance gain if the 0th device in the pair only talks to 0th
+ * devices in the remote and so on.  Unfortunately, the hypervisor is
+ * not consistent in mapping BDFs such that the natural Libfabric BDF
+ * sorting ends up with this pairing happening naturally.  So this
+ * code reorders the provider list to make that happen.  There are two
+ * pairs of cards behind each PCIe switch, and we don't actually care
+ * which is which, only that the oth device talks to the 0th device in
+ * and so on.
+ *
+ * We maintain BDF ordering in general, but do minimal reordering so
+ * that there is an alternating of the 0th pair index and then the 1st
+ * pair index, and so on.  This should work on all platforms on
+ * roadmap, but feels a little icky because it does not guarantee that
+ * the first N providers are all actually sharing resources (but it
+ * turns out for the one place where we have this sharing, that
+ * doesn't matter).  Assuming we get better at BDF ordering, this
+ * should all hopefully become a no-op in the future anyway.
+ */
 void platform_sort_rails(struct fi_info **info_list, size_t num_rails, size_t num_groups)
 {
-	struct fi_info *info_list_in = *info_list;
-	struct fi_info **sorted_info_array = (struct fi_info **)alloca(num_rails*sizeof(struct fi_info *));
-	struct fi_info *info_ptr = NULL;
-
-	if (num_rails <= 0) {
+	struct fi_info **info_array = NULL;
+	struct fi_info *info_iter = NULL;
+	size_t *vf_array = NULL;
+	struct fi_info *output_info_list = NULL;
+	struct fi_info *output_info_end = NULL;
+	size_t highest_vf_idx = 0;
+	size_t next_vf_idx = 0;
+	size_t info_count;
+
+	/* we only want to reorder if there's more than one NIC per
+	 * group (ie, per GPU).  Less than that (P4d or trainium), we
+	 * assume topo ordering is sufficient */
+	if ((num_rails / num_groups) <= 1) {
 		return;
 	}
 
-	for (size_t i = 0; i < num_rails; ++i) {
-		sorted_info_array[i] = NULL;
+	info_array = (struct fi_info **)calloc(num_rails, sizeof(struct fi_info*));
+	if (info_array == NULL) {
+		NCCL_OFI_WARN("Did not reorder arrays due to calloc failure");
+		goto cleanup;
 	}
 
-	size_t rail_map[2] = {0, 2};
+	vf_array = (size_t *)calloc(num_rails, sizeof(size_t));
+	if (vf_array == NULL) {
+		NCCL_OFI_WARN("Did not reorder arrays due to calloc failure");
+		goto cleanup;
+	}
 
-	for (size_t i = 0; i < num_rails; ++i) {
-		if (info_list_in == NULL) {
-			goto error;
+	/* copy the input list into an array so that we can more *
+	 * easily associate more data (like the vf array) with the
+	 * input and keep everything organized */
+	info_iter = *info_list;
+	info_count = 0;
+	while (info_iter != NULL) {
```

Comment:
should we have this as condition? 
`info_iter != NULL && info_count<num_rails`
alternatively, an if statement in the loop raising an error if the array index goes out-of-bounds.

We guarantee it in the way we call this function, but we could mess up in the future.

Reply:
yeah, fair.